src/ripple/app/misc/impl/ValidatorList.cpp

```diff
@@ -0,0 +1,269 @@
+    PublicKey const& identity)
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+    return ephemeral_.erase (identity);
+}
+
+bool
+ValidatorList::insertPermanentKey (
+    RippleAddress const& identity,
+    std::string const& comment)
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+
+    auto const publicKey = asPublicKey (identity);
+
+    if (ephemeral_.find (publicKey) != ephemeral_.end())
+    {
+        JLOG (j_.error) <<
+            toBase58 (TokenType::TOKEN_NODE_PUBLIC, publicKey) <<
+            ": permanent key exists in ephemeral table!";
+        return false;
+    }
+
+    return permanent_.emplace (publicKey, comment).second;
+}
+
+bool
+ValidatorList::removePermanentKey (
+    RippleAddress const& identity)
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+    return permanent_.erase (asPublicKey (identity));
+}
+
+void
+ValidatorList::for_each (
+    std::function<void(PublicKey const&, std::string const&, bool)> func) const
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+
+    for (auto const& v : permanent_)
+        func (v.first, v.second, false);
+    for (auto const& v : ephemeral_)
+        func (v.first, v.second, true);
+}
+
+//------------------------------------------------------------------------------
+static
+std::size_t
+loadValidators (
+    ValidatorList *vl,
+    std::vector<std::string> const& entries,
+    beast::Journal& j)
+{
+    static boost::regex const re (
+        "[[:space:]]*"            // skip leading whitespace
+        "([[:alnum:]]+)"          // node identity
+        "(?:"                     // begin optional comment block
+        "[[:space:]]+"            // (skip any amount of whitespace)
+        "(.*[^[:space:]]+)"       // the comment (at least non-whitespace)
+        "[[:space:]]*"            // (skip any amount of whitespace)
+        ")?"                      // end optional comment block
+    );
+
+    std::size_t count = 0;
+
+    for (auto const& n : entries)
+    {
+        boost::smatch match;
+
+        if (!boost::regex_match (n, match, re))
+        {
+            JLOG (j.error) <<
+                "Malformed entry: '" << n << "'";
+            continue;
+        }
+
+        auto const id = parseBase58<PublicKey>(
+            TokenType::TOKEN_NODE_PUBLIC, match[1]);
+
+        if (!id)
+        {
+            JLOG (j.error) <<
+                "Invalid node identity: " << match[1];
+            continue;
+        }
+
+        auto const ra = RippleAddress::createNodePublic (match[1]);
+
+        if (vl->member (ra))
+        {
+            JLOG (j.warning) <<
+                "Duplicate node identity: " << match[1];
+            continue;
+        }
+
+        if (vl->insertPermanentKey(ra, trim_whitespace (match[2])))
+            count++;
+    }
+
+    return count;
+}
+
+//------------------------------------------------------------------------------
+std::unique_ptr<ValidatorList>
+make_ValidatorList (Config const& config, beast::Journal j)
```

Comment:
I hear you re: `make_` functions not implementing business logic. If we want to reduce the use of pure virtual interfaces and separate implementations, we ought to reconsider this: The model allows us to keep the class interface tight, provided that the class can be fully configured given its existing public interface.

For another example of how I think this works nicely, look at https://github.com/nbougalis/rippled/commit/9cf1961e250ff708c3b1aaa9f9ca20550fa65990#diff-d88458c074ef19c81d06847d14619e65R31. The function doesn't return `std::unique_ptr` or `std::shared_ptr` but it is indicative of what he function does, and this eliminated an entire class.

We can go with a two-phase construction if you want, but that has its own can of worms: the class can be in a state where it's constructed but not initialized - and such a state might be meaningless.

Now, if you object to the use of the `make_` prefix, we could always choose some other name, but I think it is descriptive and accurate.